src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -950,10 +950,6 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     al, right->base.loc, right, ASR::cast_kindType::LogicalToInteger, int_type));
                 return ASR::down_cast<ASR::expr_t>(ASRUtils::make_Cast_t_value(
                     al, right->base.loc, right, ASR::cast_kindType::IntegerToComplex, left_type));
-            } else {
-                std::string rtype = ASRUtils::type_to_str_python(right_type);
-                throw SemanticError("Casting " + rtype + " to complex is not Implemented",
-                        right->base.loc);
```

Comment:
This change makes LPython return the correct message.

One thing that worries me is that this branch of the if statement structure in this function does not do anything after this change. Is that a good design of `cast_helper`? 

It seems that `cast_helper` gives an error message right away if things cannot be cast. In which case, maybe we should just reformulate what this error message says. Instead of saying "not implemented", we should say "cannot cast c32 to string". 

I do like the current error message (in this PR, as seen in the tests). Do we effectively have two places in LPython that return the error? In `cast_helper` and then later as well?

Should we just have one place, say `cast_helper`, that checks for errors, and then later we assume they are correct?

Reply:
Well IMO,

> Is that a good design of `cast_helper`?

`cast_helper` is not abstract enough. Instead of a long `if-else-if` ladder I would prefer a simple rule map based with a tuple of type as keys and the desired cast as the value. Just pass in the types to the rule map and get your cast type. But anyways doesn't matter much now as its already implemented so let's keep using it.

> Instead of saying "not implemented", we should say "cannot cast c32 to string".

Agreed. The second message is more transparent and helps in debugging.

> Should we just have one place, say cast_helper, that checks for errors,

In an abstract sense, yes there should be only one place which checks for type compatibilities and return errors. If no errors then assume correctness in rest of your implementation.

Reply:
@faze-geek Why have you removed this `SemanticError` anyways?

Reply:
The current `SemanticError` is inconsistent (as pointed in #673) so I first I was changing the error inside the `else` itself . But then by `--show-stacktrace` I realised that by removing these lines , the `type mismatch` errors for 1. `str + int`, 2. `str + float ` and 3. `str + complex` were getting generated from the same lines in the code ( which I thought is better than error generated from different lines)

Reply:
I thought this would be straightforward ,I was unaware of `cast_helper`.